vision.py

```diff
@@ -117,26 +119,64 @@ def find_loading_bay(self, frame: np.ndarray):
+
+        cnts, _ = cv2.findContours(frame, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)
         if len(cnts) >= 1:
             acceptable_cnts = []
+            # Check if the found contour is possibly a target
             for current_contour in enumerate(cnts):
                 area = cv2.contourArea(current_contour[1])
-                box = cv2.boundingRect(current_contour[1])
-                hull_area = cv2.contourArea(cv2.convexHull(current_contour[1]))
-                if (
-                    area > MIN_CONTOUR_AREA
-                    and area / hull_area > POWER_PORT_AREA_RATIO
-                    and box[2] > box[3]
-                ):
-                    acceptable_cnts.append(current_contour[1])
+                if PP_MAX_CONTOUR_AREA > area > PP_MIN_CONTOUR_AREA:
+                    box = cv2.boundingRect(current_contour[1])
+                    # Convex hull gives the bounding polygon of the contour with no
+                    # interior angles greater than 180deg
+                    hull = cv2.convexHull(current_contour[1])
+                    hull_area = cv2.contourArea(hull)
+                    # If the contour takes up more than X% of the Hull and
+                    # width greater than height
+                    if (
+                        PP_MAX_AREA_RATIO > area / hull_area > PP_MIN_AREA_RATIO
+                        and box[2] > box[3]
+                    ):
+                        # print(box) # X,Y,W,H
+                        # print("P %.2f, Area %d, Hull, %d" % (area / hull_area, area, hull_area))
+                        acceptable_cnts.append(current_contour[1])
+                        hullList.append(hull)
+
+            # ***This section of code displays the possible targets***
+            for i in range(len(acceptable_cnts)):
+                color_G = (0, 255, 0)
+                color_B = (255, 0, 0)
+                cv2.drawContours(self.display, acceptable_cnts, i, color_G)
+                cv2.drawContours(self.display, hullList, i, color_B)
 
             if acceptable_cnts:
-                power_port_contour = max(
-                    acceptable_cnts, key=lambda x: cv2.contourArea(x)
-                )
+                if len(acceptable_cnts) > 1:
+                    # Pick the largest found 'power port'
+                    power_port_contour = max(
+                        acceptable_cnts, key=lambda x: cv2.contourArea(x)
+                    )
+                else:
+                    power_port_contour = acceptable_cnts[0]
                 power_port_points = get_corners_from_contour(power_port_contour)
                 # x, y, w, h = cv2.boundingRect(power_port_contour)
+                for i in range(4):
+                    cv2.circle(
+                        self.display, tuple(power_port_points[i][0]), 3, (0, 0, 255)
+                    )
+                cv2.imshow("Display", self.display)
```

Comment:
Oops. This one isn't in a test, and it doesn't have a timeout. This will hang the robot.